### PR TITLE
[CF-249] Fix floating IPs migration

### PR DESCRIPTION
**Pull request by miarmak**
Change the way of networks comparison. Collect all subnets hashes of
each network and use it for more accurate networks comparison
(identify which network on DST matches with the same network from SRC).

### CODE REVIEW

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.os.network.neutron
cloudferrylib/os/network/neutron.py:184: [W0221(arguments-differ), 
NeutronNetwork.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/network/neutron.py:289: [W0221(arguments-differ), 
NeutronNetwork.create_port] Arguments number differs from overridden 
'create_port' method
cloudferrylib/os/network/neutron.py:821: [W0221(arguments-differ), 
NeutronNetwork.get_security_groups] Arguments number differs from overridden 
'get_security_groups' method

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferrylib/os/network/neutron.py [hunk 1]
```diff
@@ -1485,13 +1480,51 @@ def get_resource_hash(neutron_resource, *args):
...
+        :return dict: Network object from DST, that matches with the same
+                      network from SRC.
+        """
+
+        for net in existing_networks:
+            if (net['res_hash'] == src_net['res_hash'] and
+                    net['subnets_hash'] == src_net['subnets_hash']):
+                return net
+
+        return None
```
Comment (roman-verchikov) [inline cloudferrylib/os/network/neutron.py]: python returns `None` by default, doing that explicitly is not really needed.

Comment (miarmak) [inline cloudferrylib/os/network/neutron.py]: I specifically used it here for better visibility.

File: cloudferrylib/os/network/neutron.py [hunk 2]
```diff
@@ -1485,13 +1480,51 @@ def get_resource_hash(neutron_resource, *args):
...
         else:
-            net_hash = self.get_res_hash_by_id(src_nets, src_net_id)
-            dst_net_id = self.get_res_by_hash(dst_nets, net_hash)['id']
+            src_net = get_network_from_list_by_id(src_net_id, src_nets)
+            dst_net = self.get_dst_net_by_src_net(dst_nets, src_net)
+
+            if not dst_net:
+                raise exception.AbortMigrationError(
+                    "No any suitable external network has been found. "
+                    "Aborting migration...")
```
Comment (roman-verchikov) [inline cloudferrylib/os/network/neutron.py]: Just return `None` here, this case is already handled by calling code

File: cloudferrylib/os/network/neutron.py (hunk 1 above)
Comment (miarmak) [inline cloudferrylib/os/network/neutron.py]: Done

File: cloudferrylib/os/network/neutron.py (hunk 2 above)
Comment (miarmak) [inline cloudferrylib/os/network/neutron.py]: Done

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.os.network.neutron
cloudferrylib/os/network/neutron.py:184: [W0221(arguments-differ), 
NeutronNetwork.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/network/neutron.py:289: [W0221(arguments-differ), 
NeutronNetwork.create_port] Arguments number differs from overridden 
'create_port' method
cloudferrylib/os/network/neutron.py:821: [W0221(arguments-differ), 
NeutronNetwork.get_security_groups] Arguments number differs from overridden 
'get_security_groups' method

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by roman-verchikov** [comment]
+1, looks good to me

**Comment by roman-verchikov** [comment]
please rebase

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.os.network.neutron
cloudferrylib/os/network/neutron.py:188: [W0221(arguments-differ), 
NeutronNetwork.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/network/neutron.py:294: [W0221(arguments-differ), 
NeutronNetwork.create_port] Arguments number differs from overridden 
'create_port' method
cloudferrylib/os/network/neutron.py:347: [W0221(arguments-differ), 
NeutronNetwork.check_existing_port] Arguments number differs from overridden 
'check_existing_port' method
cloudferrylib/os/network/neutron.py:839: [W0221(arguments-differ), 
NeutronNetwork.get_security_groups] Arguments number differs from overridden 
'get_security_groups' method

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferrylib/os/actions/check_networks.py [hunk 3]
```diff
@@ -462,7 +462,7 @@ def get_invalid_ext_net_ids(self, dst_info, ext_net_map):
 class Network(object):
     def __init__(self, info):
         self.info = info
-        self.subnets_hash = set()
+        self.subnets_hash = info['subnets_hash']
```
Comment (SVilgelm) [inline cloudferrylib/os/actions/check_networks.py]: Are you sure that 'subnets_hash` is always present in the `info`?

Comment (miarmak) [inline cloudferrylib/os/actions/check_networks.py]: Yes. If network doe not have any subnet, it will be just empty set.

Comment (SVilgelm) [inline cloudferrylib/os/actions/check_networks.py]: Ok. thanks

**Comment by miarmak** [comment]
retest this please

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.os.network.neutron
cloudferrylib/os/network/neutron.py:188: [W0221(arguments-differ), 
NeutronNetwork.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/network/neutron.py:294: [W0221(arguments-differ), 
NeutronNetwork.create_port] Arguments number differs from overridden 
'create_port' method
cloudferrylib/os/network/neutron.py:347: [W0221(arguments-differ), 
NeutronNetwork.check_existing_port] Arguments number differs from overridden 
'check_existing_port' method
cloudferrylib/os/network/neutron.py:839: [W0221(arguments-differ), 
NeutronNetwork.get_security_groups] Arguments number differs from overridden 
'get_security_groups' method

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by roman-verchikov** [comment]
+1, needs to be tested in staging

**Comment by yosharma** [comment]
+1, Tested on dfw3 environment.

**Comment by roman-verchikov** [comment]
can be merged

**Pull request merged**